developer_guides/firmware/async_messaging_best_practices.rst

```diff
@@ -0,0 +1,651 @@
+     if(error != 0) {
+       //error handling
+     }
+
+     //registering consumer
+     message_context.error = 0;
+     message_context.request_queue = &request_queue;
+     error = am_service_register_consumer(message_type_id, GetcomponentID(), GetInstanceID(), callback, &message_context);
+     if(error != 0) {
+       //error handling
+     }
+   }
+
+   error_code componentA::ProcessData(size_t max_output_data_size, uint32_t* custom_error_code) {
+      while(queue_size(message_context.request_queue) > 0) {
+         if(message_context.error != 0) {
+           //error handling
+         }
+
+         request_t *r = queue_pop(message_context.request_queue);
+         //handling of async message
+         process_request(r);
+      }
+
+      //regular processing
+   }
+
+.. _processing_in_a_consumer_callback:
+
+Processing in a consumer callback
+*********************************
+
+The consumer callbacks must not do any heavy processing. Best if the callback
+only saves information and a component will process it in the next processing
+slot.
+
+-  The asynchronous message callbacks are called in the context of a message
+   producer or in the context of IDC task.
+-  If a heavy processing is done in a message, then a producer MPCS or IDC task
+   budget needs to take into account this heavy processing. While it may be
+   possible to take into account additional budget for simple cases, it is
+   impossible to accommodate MCPS requirements in the general case (i.e. large
+   number of consumers with heavy processing).
+
+The example described in :ref:`message_handling` is also applicable for this issue.
+
+.. _message_filtering_mechanism:
+
+Message Filtering Mechanism
+***************************
+
+1. If a consumer is interested in async messages from one particular producer,
+   it needs to register for that particular producer using component ID and
+   instance ID. A component ID and instance ID of producer should be passed as
+   ``LARGE_CONFIG_SET`` parameter to the consumer.
+2. If a producer wants to target a specific component instance, then it should
+   use send function that is parametrized with component ID and instance ID.
+
+Passing Pointers In Asynchronous Messages
+*****************************************
+
+The asynchronous messages do not prevent to pass pointers between components. It
+seems like a good idea at the beginning when a developer wants to return a
+result. While it is simple solution, it may lead into the following issues:
+
+-  there can be more than one consumer of a message and each of them needs to
+   have an allocated slot in the memory,
+-  the firmware framework cannot guarantee that component memory will be still
+   available when async message is processed i.e. a component can be subject of
+   firmware paging and the firmware infrastructure can decide to evict component
+   memory,
+
+The recommended method is to not pass pointers in the asynchronous messages.
```

Comment:
ack - this could be enforced ?

Reply:
I would say that AMS should be implemented in a generic way, without enforcement. However the planned usage is going to be limited to just sending single event information. If there will be need to pass a pointer for some reason then it should be individually review and decided whether it is a right approach.